From bb155ed5ce9230229fa488b74e800968d724d172 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Wed, 28 Nov 2012 14:28:15 +0900
Subject: [PATCH 29/39] mpeg4dec: specify the video stream type as packed
 bitstream

REL OMXIL can be set the video stream type, which is one of the
extended parameters. This setting is necessary to decode mpeg4 streams
with packed bitstreams. Non-packed bitstreams can be decoded regardless
of the value of this setting, so it is safe to simply turn it on for
all mpeg4 streams.
This patch overrides base omx_setup() method of mpeg4dec
in order to set this param.
---
 omx/gstomx_mpeg4dec.c |   31 +++++++++++++++++++++++++++++++
 omx/gstomx_mpeg4dec.h |    2 ++
 2 files changed, 33 insertions(+)

diff --git a/omx/gstomx_mpeg4dec.c b/omx/gstomx_mpeg4dec.c
index 1286f65..7e5cf73 100644
--- a/omx/gstomx_mpeg4dec.c
+++ b/omx/gstomx_mpeg4dec.c
@@ -19,8 +19,10 @@
  *
  */
 
+#include <string.h>             /* for memset */
 #include "gstomx_mpeg4dec.h"
 #include "gstomx.h"
+#include "OMXR_Extension.h"
 
 GSTOMX_BOILERPLATE (GstOmxMpeg4Dec, gst_omx_mpeg4dec, GstOmxBaseVideoDec,
     GST_OMX_BASE_VIDEODEC_TYPE);
@@ -47,6 +49,28 @@ type_base_init (gpointer g_class)
 }
 
 static void
+omx_setup (GstOmxBaseFilter * omx_base)
+{
+  GstOmxMpeg4Dec *omx_mpeg4dec;
+  GOmxCore *gomx;
+  OMXR_MC_VIDEO_PARAM_STREAM_STORE_UNITTYPE store_type;
+
+  omx_mpeg4dec = GST_OMX_MPEG4DEC (omx_base);
+  gomx = (GOmxCore *) omx_base->gomx;
+
+  memset (&store_type, 0, sizeof (store_type));
+  store_type.nSize = sizeof (store_type);
+
+  OMX_GetParameter (gomx->omx_handle, OMXR_MC_IndexParamVideoStreamStoreUnit,
+      &store_type);
+  store_type.eStoreUnit = OMXR_MC_VIDEO_StoreUnitPackedBitstream;
+  OMX_SetParameter (gomx->omx_handle, OMXR_MC_IndexParamVideoStreamStoreUnit,
+      &store_type);
+
+  omx_mpeg4dec->base_omx_setup (omx_base);
+}
+
+static void
 type_class_init (gpointer g_class, gpointer class_data)
 {
 }
@@ -55,8 +79,15 @@ static void
 type_instance_init (GTypeInstance * instance, gpointer g_class)
 {
   GstOmxBaseVideoDec *omx_base;
+  GstOmxBaseFilter *omx_base_filter;
+  GstOmxMpeg4Dec *omx_mpeg4dec;
 
   omx_base = GST_OMX_BASE_VIDEODEC (instance);
+  omx_base_filter = GST_OMX_BASE_FILTER (instance);
+  omx_mpeg4dec = GST_OMX_MPEG4DEC (instance);
 
   omx_base->compression_format = OMX_VIDEO_CodingMPEG4;
+
+  omx_mpeg4dec->base_omx_setup = omx_base_filter->omx_setup;
+  omx_base_filter->omx_setup = omx_setup;
 }
diff --git a/omx/gstomx_mpeg4dec.h b/omx/gstomx_mpeg4dec.h
index 1b5cce8..46f5d60 100644
--- a/omx/gstomx_mpeg4dec.h
+++ b/omx/gstomx_mpeg4dec.h
@@ -35,6 +35,8 @@ typedef struct GstOmxMpeg4DecClass GstOmxMpeg4DecClass;
 struct GstOmxMpeg4Dec
 {
   GstOmxBaseVideoDec omx_base;
+
+  GstOmxBaseFilterCb base_omx_setup;
 };
 
 struct GstOmxMpeg4DecClass
-- 
1.7.9.5

